### PR TITLE
Fix serious FSDirectory use-after-Close bug (missing fsync)

### PR DESCRIPTION
**Pull request by pdillinger**
Summary:
TL;DR: due to a recent change, if you drop a column family,
often that DB will no longer fsync after writing new SST files
to remaining or new column families, which could lead to data
loss on power loss.

More bug detail:
The intent of #10049 was to Close FSDirectory objects at
DB::Close time rather than waiting for DB object destruction.
Unfortunately, it also closes shared FSDirectory objects on
DropColumnFamily (& destroy remaining handles), which can lead
to use-after-Close on FSDirectory shared with remaining column
families. Those "uses" are only Fsyncs (or redundant Closes). In
the default Posix filesystem, an Fsync on a closed FSDirectory is a
quiet no-op. Consequently (under most configurations), if you drop
a column family, that DB will no longer fsync after writing new SST
files to column families sharing the same directory (true under most
configurations).

More fix detail:
Basically, this removes unnecessary Close ops on destroying
ColumnFamilyData. We let `shared_ptr` take care of calling the
destructor at the right time. If the intent was to require Close be
called before destroying FSDirectory, that was not made clear by the
author of FileSystem and was not at all enforced by #10049, which
could have added `assert(fd_ == -1)` to `~PosixDirectory()` but did
not. To keep this fix simple, we relax the unit test for #10049 to allow
timely destruction of FSDirectory to suffice as Close (in
CountedFileSystem). Added a TODO to revisit that.

Also in this PR:
* Added a TODO to share FSDirectory instances between DB and its column
families. (Already shared among column families.)
* Made DB::Close attempt to close all its open FSDirectory objects even
if there is a failure in closing one. Also code clean-up around this
logic.

Test Plan: add an assert to check for use-after-Close. With that
existing tests can detect the misuse. With fix, tests pass (except noted
relaxing of unit test for #10049)

### CODE REVIEW

File: db/column_family.cc [hunk 1]
```diff
@@ -699,19 +699,6 @@ ColumnFamilyData::~ColumnFamilyData() {
           id_, name_.c_str());
     }
   }
-
-  if (data_dirs_.size()) {  // Explicitly close data directories
```
Comment (littlepig2013) [inline db/column_family.cc:703]: Thanks a lot for catching this bug! @pdillinger I wonder if maintaining an unordered_map between data directories and the ref counter could solve this? Here we just decrease the ref counter for each data directory and close the data directory when the ref counter is 0? Is this feasible?

File: utilities/counted_fs.cc [hunk 2]
```diff
@@ -242,6 +244,14 @@ class CountedDirectory : public FSDirectoryWrapper {
     }
     return rv;
   }
+
+  ~CountedDirectory() {
+    if (!closed_) {
+      // TODO: fix DB+CF code to use explicit Close, not rely on destructor
```
Comment (littlepig2013) [inline utilities/counted_fs.cc:250]: If the ref counter is feasible, we may already fix this TODO and we can remove these?

File: db/column_family.cc (hunk 1 above)
Comment (pdillinger) [reply db/column_family.cc:703]: I should have mentioned: this PR is intended for patching previous releases. For that we want a kind of minimum sufficient fix.

And I authored this PR myself because I got most of the way here to help me understand the problem. Proper treatment of the refcounting stuff should (IMHO) incorporate both of my TODOs, so it's not a simple change and I don't intend to follow-up on it myself any time soon.

I'm not even really convinced it's that useful to ensure we always call Close. The practical effect of the complete ambiguity in the original API definition is that implementors need to effect a Close on destruction if it hasn't been called already. What do we gain by enforcing a stronger constraint on ourselves? And I'm not sure what kind of useful failure FSDirectory::Close would return to us in any environment. There can't (at present) be any pending buffered operations that might have failed in the background. If the intent was to future-proof the API in case such operations were added in the future, then Close should have been marked and enforced as required. At present, we don't even enforce Close on DB, which can return background errors like "your entire DB disappeared! It's probably not going to be here when you try to re-open!"

**Review by littlepig2013 — APPROVED** [review]
LGTM

File: db/column_family.cc (hunk 1 above)
Comment (littlepig2013) [reply db/column_family.cc:703]: I may follow up this later to refine the implementation and your point makes sense to me. I will take them into account and discuss with you later if I have something else proposed. As we are aiming for sufficient minimum changes, this PR looks good to me.

File: db/column_family.cc [hunk 3]
```diff
@@ -699,19 +699,6 @@ ColumnFamilyData::~ColumnFamilyData() {
...
-    Status s = Status::OK();
-    for (auto& data_dir_ptr : data_dirs_) {
```
Comment (ajkr) [inline db/column_family.cc:705]: Sorry I didn't realize `ColumnFamilyData::data_dirs_` is sometimes shared. Looks like it wasn't initially in #3102, then became sometimes shared in #6358 (interestingly it was still effectively private for CFDs created via `CreateColumnFamily()`). I agree it should be shared, preferably always (including shared with `DBImpl::data_dirs_`).

I bet we'd have noticed in review if not for the use of `auto` here.

**Review by ajkr — APPROVED** [review]
LGTM, thanks!

File: db/column_family.cc (hunk 1 above)
Comment (ajkr) [reply db/column_family.cc:703]: > I'm not even really convinced it's that useful to ensure we always call Close. The practical effect of the complete ambiguity in the original API definition is that implementors need to effect a Close on destruction if it hasn't been called already. What do we gain by enforcing a stronger constraint on ourselves?

It's a good question. The original goal of #10049 was to release file descriptors when `DB::Close()` is called. It looks like it could've been achieved by clearing `DBImpl::data_dirs_` and resetting `DBImpl::{db,wal}_dir_` during `DB::Close()`.
